wait_all.go

```diff
@@ -0,0 +1,86 @@
+package asynctask
+
+import (
+	"context"
+	"fmt"
+	"sync"
+)
+
+// WaitAllOptions defines options for WaitAll function
+type WaitAllOptions struct {
+	// FailFast set to true will indicate WaitAll to return on first error it sees.
+	FailFast bool
+}
+
+// WaitAll block current thread til all task finished.
+// first error from any tasks passed in will be returned.
+func WaitAll(ctx context.Context, options *WaitAllOptions, tasks ...*TaskStatus) error {
+	tasksCount := len(tasks)
+	errorCh := make(chan error, tasksCount)
+	errorChClosed := false
+	defer close(errorCh)
+	mutex := sync.Mutex{}
+
+	for _, tsk := range tasks {
+		go func(tsk *TaskStatus) {
+			_, err := tsk.Wait(ctx)
+
+			// why mutex?
+			// if all tasks start using same context (unittest is good example)
+			// and that context got canceled, all task fail at same time.
+			// first one went in and close the channel, while another one already went through gate check.
+			// raise a panic with send to closed channel.
+			mutex.Lock()
+			defer mutex.Unlock()
+			if !errorChClosed {
+				errorCh <- err
+			}
+		}(tsk)
+	}
+
+	runningTasks := tasksCount
+	var errList []error
+	for {
+		select {
+		case err := <-errorCh:
+			runningTasks--
+			if err != nil {
+				// return immediately after receive first error.
+				if options.FailFast {
+					closeChannel(errorCh, &mutex, &errorChClosed)
+					return err
+				}
+
+				errList = append(errList, err)
+			}
+		case <-ctx.Done():
+			closeChannel(errorCh, &mutex, &errorChClosed)
+			return fmt.Errorf("WaitAll context canceled: %w", ctx.Err())
+		}
+
+		// are we finished yet?
+		if runningTasks == 0 {
```

Comment:
you already have a `defer close(errorCh)`, so you are guaranteed that the channel gets closed when you get out of WaitAll().
Wouldn't that avoid the mutex, because you only close the channel after all tasks are handled?